### PR TITLE
c7n-left - graph traversal filter

### PR DESCRIPTION
**Pull request by kapilt**
support nested graph traversal going from a terraform block to any other set of referenced blocks, with multi 
attribute matching at destination.

ideally should use list-item under hood in future, but schema is future compatible to that.

also unclear on naming, `traverse` vs `link`

### CODE REVIEW

File: tools/c7n_left/c7n_left/providers/terraform/graph.py [hunk 1]
```diff
@@ -36,9 +39,65 @@ def get_resources_by_type(self, types=()):
...
+        return self.resolver.resolve_refs(resource, (target_type,))
+
+
+class Resolver:
+    def __init__(self):
+        self._id_map = {}
+        self._ref_map = {}
+
+    @staticmethod
+    def is_id_ref(v):
```
Comment (sontek) [inline tools/c7n_left/c7n_left/providers/terraform/graph.py:60]: This function isn't necessary once we merge:

https://github.com/cloud-custodian/tfparse/pull/58

It adds `__ref__` to the value so its clear which ones are actual references.

File: tools/c7n_left/c7n_left/core.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 from c7n.provider import Provider, clouds
 from c7n.policy import PolicyExecutionMode
 
+from .filters import Link
```
Comment (sontek) [inline tools/c7n_left/c7n_left/core.py:15]: Looks like you forgot to commit this code

Comment (kapilt) [reply tools/c7n_left/c7n_left/core.py:15]: doh..

File: tools/c7n_left/c7n_left/providers/terraform/graph.py (hunk 1 above)
Comment (kapilt) [reply tools/c7n_left/c7n_left/providers/terraform/graph.py:60]: cool, that's a nice change up for tfparse as well, and its nice to get away from implicit string checks here.

**Comment by sontek** [comment]
@kapilt This looks great now.  Could you add a test that shows how to use jmespath to filter a value on a sub resource with this?   I'm thinking the s3 <--> server side encryption configuration relationship.   So for example if you have this terraform:

```
resource "aws_s3_bucket" "example_d" {
  bucket = "c7n-aws-s3-encryption-audit-test-d"
  acl    = "private"
}

resource "aws_s3_bucket_server_side_encryption_configuration" "example_d_sse" {
  bucket = aws_s3_bucket.example_d.bucket

  rule {
    apply_server_side_encryption_by_default {
      sse_algorithm     = "AES256"
    }
  }
}
```

and you wanted to make sure they are using `aws:kms` and not `AES256`, what would the filter look like for that?  i.e how do you get to `rule.apply_server_side_encryption_by_default.sse_algorithm` from the s3 bucket?

**Comment by kapilt** [comment]
agreed the filter needs some more documentation examples, wrt to the specific question (and untested)

```yaml
policies:
  - name: check-bucket-encryption
     resource: aws_s3_bucket
     filters:
       - type: link
          resources: aws_s3_bucket_server_side_encryption_configuration
          attrs:
             - rule.apply_server_side_encryption_by_default.sse_algorithm: AES256
```

still not digging the name `link`, what do you think of `traverse` or `resolve`

**Comment by sontek** [comment]
Works, here is a test for it:

https://github.com/kapilt/cloud-custodian/pull/14

**Comment by sontek** [comment]
For the naming, I almost don't think `link`, `traverse`, or `resolve` are not the right name because its more of an implementation detail.   Realistically this is just a generic filtering based on the datastructure generated from the provider. It is probably the only filter most people will need to use with these providers.

I don't have a better name though.   If I had to choose I think `traverse` would be what I'd go with though

**Review by sontek — APPROVED** [review]

**Comment by sontek** [comment]
@kapilt Is there a way to ask for `absent` with this?   Like:

```
  - name: test-sqs-is-encrypted
    resource: terraform.aws_sqs_queue
    filters:
      - type: link
        resources:
          - aws_sqs_queue
        attrs:
          - sqs_managed_sse_enabled: absent


```

**Comment by sontek** [comment]
@kapilt Found one bug we might want to fix but can also be a follow up, if we want to use `boolean` filters:

```
Traceback (most recent call last):
  File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/bin/c7n-left", line 6, in <module>
    sys.exit(cli())
  File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/cli.py", line 51, in run
    runner.run()
  File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/core.py", line 62, in run
    result_set = self.run_policy(p, graph, resources, event)
  File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/core.py", line 70, in run_policy
    return policy.push(event)
  File "/Users/john/code/cloud-custodian/c7n/policy.py", line 1288, in push
    return mode.run(event, lambda_ctx)
  File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/core.py", line 99, in run
    resources = self.manager.filter_resources(resources, event)
  File "/Users/john/code/cloud-custodian/c7n/manager.py", line 111, in filter_resources
    resources = f.process(resources, event)
  File "/Users/john/code/cloud-custodian/c7n/filters/core.py", line 337, in process
    return self.process_set(resources, event)
  File "/Users/john/code/cloud-custodian/c7n/filters/core.py", line 348, in process_set
    rtype_id = self.get_resource_type_id()
  File "/Users/john/code/cloud-custodian/c7n/filters/core.py", line 316, in get_resource_type_id
    resource_type = self.manager.get_model()
  File "/Users/john/code/cloud-custodian/c7n/manager.py", line 124, in get_model
    return self.query.resolve(self.resource_type)
AttributeError: 'TerraformResourceManager' object has no attribute 'query'
```

If you do a filter like:

```
  - name: test-sqs-is-encrypted
    resource: terraform.aws_sqs_queue
    filters:
      - or:
        - type: value
          key: sqs_managed_sse_enabled
          value: absent

        - type: value
          key: sqs_managed_sse_enabled
          value: false
```

**Comment by kapilt** [comment]
> @kapilt Found one bug we might want to fix but can also be a follow up, if we want to use `boolean` filters:
> 
> ```
> Traceback (most recent call last):
>   File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/bin/c7n-left", line 6, in <module>
>     sys.exit(cli())
>   File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
>     return self.main(*args, **kwargs)
>   File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1055, in main
>     rv = self.invoke(ctx)
>   File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/Users/john/Library/Caches/pypoetry/virtualenvs/c7n-nOD_wM_f-py3.10/lib/python3.10/site-packages/click/core.py", line 760, in invoke
>     return __callback(*args, **kwargs)
>   File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/cli.py", line 51, in run
>     runner.run()
>   File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/core.py", line 62, in run
>     result_set = self.run_policy(p, graph, resources, event)
>   File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/core.py", line 70, in run_policy
>     return policy.push(event)
>   File "/Users/john/code/cloud-custodian/c7n/policy.py", line 1288, in push
>     return mode.run(event, lambda_ctx)
>   File "/Users/john/code/cloud-custodian/tools/c7n_left/c7n_left/core.py", line 99, in run
>     resources = self.manager.filter_resources(resources, event)
>   File "/Users/john/code/cloud-custodian/c7n/manager.py", line 111, in filter_resources
>     resources = f.process(resources, event)
>   File "/Users/john/code/cloud-custodian/c7n/filters/core.py", line 337, in process
>     return self.process_set(resources, event)
>   File "/Users/john/code/cloud-custodian/c7n/filters/core.py", line 348, in process_set
>     rtype_id = self.get_resource_type_id()
>   File "/Users/john/code/cloud-custodian/c7n/filters/core.py", line 316, in get_resource_type_id
>     resource_type = self.manager.get_model()
>   File "/Users/john/code/cloud-custodian/c7n/manager.py", line 124, in get_model
>     return self.query.resolve(self.resource_type)
> AttributeError: 'TerraformResourceManager' object has no attribute 'query'
> ```
> 
> If you do a filter like:
> 
> ```
>   - name: test-sqs-is-encrypted
>     resource: terraform.aws_sqs_queue
>     filters:
>       - or:
>         - type: value
>           key: sqs_managed_sse_enabled
>           value: absent
> 
>         - type: value
>           key: sqs_managed_sse_enabled
>           value: false
> ```

fixed, also use value: present to cover both cases more simply without the `or`

**Comment by kapilt** [comment]
> Works, here is a test for it:
> 
> [kapilt#14](https://github.com/kapilt/cloud-custodian/pull/14)

actually its more nuanced then that, on the policy as this policy would not detect something without encryption, so actually wants to be wrapped as a positive assertion of kms encrypt with a not: block so it matches any form that doesn't match exactly. also this isn't really worth testing (jmespath key) since its deferred on implementation to value filter, which is a workhorse. it does however want a docstring example that we can use later for the reference docs.

**Comment by kapilt** [comment]
> So I want to link the two with `sourceQueueArns` and require `sqs_managed_sse_enabled` to be true (not absent)

its unclear what the actual intent here is, if you have an sqs encryption policy your going to apply across sqs queue, irrespective of link/traversal between them.

**Comment by kapilt** [comment]
its also worth noting you can use count: 0 in the filter for negative assertions instead of wrapping with not

**Pull request merged**